notion-core/src/main/scala/com/ambiata/notion/core/LineReducer.scala

```diff
@@ -0,0 +1,104 @@
+package com.ambiata.notion
+package core
+
+import java.security.MessageDigest
+
+import org.apache.commons.codec.binary.Base64
+
+import scala.collection.mutable.Queue
+import scalaz.{Foldable, Monoid}
+
+/**
+ * "Reducer-like" class to accumulate values while
+ * reading a file and transforming them to a final value
+ * when the whole file has been read
+ */
+trait LineReducer[T] {
+  type S
+  
+  /** initial value */
+  def init: S
+
+  /** reducer function using the current line and the current state S */
+  def reduce(line: String, s: S): S
+
+  /** final value from the last state */
+  def finalise(s: S): T
+}
+
+object LineReducer {
+
+  /** zip two line reducers together - aka parallel composition */
+  implicit class ZippedLineReducer[T1](r1: LineReducer[T1]) {
+    def zip[T2](r2: LineReducer[T2]): LineReducer[(T1, T2)] = new LineReducer[(T1, T2)] {
+      type S = (r1.S, r2.S)
+
+      def init: S =
+        (r1.init, r2.init)
+
+      def reduce(line: String, s: S): S =
+        (r1.reduce(line, s._1), r2.reduce(line, s._2))
+
+      def finalise(s: S): (T1, T2) =
+        (r1.finalise(s._1), r2.finalise(s._2))
+    }
+  }
+
+  /** create a reducer from a Monoid and a function transforming each line to a value T */
+  def fromMonoid[M : Monoid](f: String => M) = new LineReducer[M] {
+    type S = M
+    
+    def init: S =
+      Monoid[S].zero
+
+    def reduce(line: String, s: S): S =
+      Monoid[S].append(f(line), s)
+
+    def finalise(s: S): S =
+      s
+  }
+
+  /** line reducer counting lines */
+  def linesNumber: LineReducer[Int] =
+    LineReducer.fromMonoid[Int](_ => 1)(scalaz.std.anyVal.intInstance)
+
+  /** line reducer computing a SHA1 */
+  def sha1: LineReducer[String] = new LineReducer[String] {
+    type S = MessageDigest
+
+    val init: MessageDigest =
+      MessageDigest.getInstance("SHA-1")
+
+    def reduce(line: String, md: MessageDigest): MessageDigest = {
+      md.update((line+"\n").getBytes)
```

Comment:
Also, the "\n" isn't great. It means `sha1sum` on the command line could give you a different answer which isn't ok - because it is likely that that is what would be used to cross check. I am not entirely sure you can do this on a line by line basis and get the right answer.